packages/integrations/mdx/src/index.ts

```diff
@@ -68,34 +69,53 @@ export default function mdx(mdxOptions: MdxOptions = {}): AstroIntegration {
 					},
 				]);
 
+				const mdxCompilerOptions: MdxRollupPluginOptions = {
+					remarkPlugins,
+					rehypePlugins,
+					jsx: true,
+					jsxImportSource: 'astro',
+					// Note: disable `.md` support
+					format: 'mdx',
+					mdExtensions: [],
+				}
+
 				updateConfig({
 					vite: {
 						plugins: [
 							{
 								enforce: 'pre',
-								...mdxPlugin({
-									remarkPlugins,
-									rehypePlugins,
-									jsx: true,
-									jsxImportSource: 'astro',
-									// Note: disable `.md` support
-									format: 'mdx',
-									mdExtensions: [],
-								}),
+								...mdxPlugin(mdxCompilerOptions),
+								// Override transform to inject layouts before MDX compilation
+								async transform(code: string, id: string) {
+									if (!id.endsWith('.mdx')) return;
+									
+									const { data: frontmatter } = matter(code);
+									if (frontmatter.layout) {
+										const { layout, ...content } = frontmatter;
+										code += `\nexport default async function({ children }) {\nconst Layout = (await import(${
+											JSON.stringify(frontmatter.layout)
+										})).default;\nreturn <Layout content={${
+											JSON.stringify(content)
+										}}>{children}</Layout> }`
+									}
+
+									const { value: compiled } = await compile(code, mdxCompilerOptions);
```

Comment:
does this return sourcemaps? If so we should return those as well.

Reply:
I'll do you one better! Instead of calling the compiler raw, I'm now calling the rollup plugin's `transform` function directly now. Should mean we get everything MDX gives us out-of-the-box 👍